keps/sig-api-machinery/2876-crd-validation-expression-language/README.md

```diff
@@ -283,14 +283,31 @@ like the `all` macro, e.g. `self.all(listItem, <predicate>)` or `self.all(mapKey
     Strategy](https://kubernetes.io/docs/reference/using-api/server-side-apply/#merge-strategy) for
     details.
   - The use of "old" is congruent with how `AdmissionReview` identifies the existing object as
-    `oldObject`. To avoid name collisions `old<propertyName>` will be treated the same as a CEL
+    `oldSelf`. To avoid name collisions `old<propertyName>` will be treated the same as a CEL
     keyword for escaping purposes (see below).
   - xref [analysis of possible interactions with immutability and
     validation](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/1101-immutable-fields#openapi-extension-x-kubernetes-immutable).
 
-- If a object property collides with a CEL keyword (see RESERVED in [CEL Syntax](https://github.com/google/cel-spec/blob/master/doc/langdef.md#syntax)),
+- If a object property name is a CEL keyword (see RESERVED in [CEL Syntax](https://github.com/google/cel-spec/blob/master/doc/langdef.md#syntax)),
   it will be escaped by prepending a _ prefix. To prevent this from causing a subsequent collision, all properties with a `_` prefix will always be
   prefixed by `__` (generally, N+1 the existing number of `_`s).
+  
+- If a object property name is a CEL language identifier (`int`, `uint`, `double`, `bool`, `string`,
+  `bytes`, `list`, `map`, `null_type`, `type`, see [CEL language
+  identifiers](https://github.com/google/cel-spec/blob/master/doc/langdef.md#values)) it is not
+  accessible as a root variable and must be accessed via `self`, .e.g. `seft.int`.
+
+- Rules may be written at the root of an object, and may make field selection into any fields
+  declared in the OpenAPIv3 schema of the CRD. This includes selection of fields in both the `spec`
+  and `status` in the same expression, e.g. `status.quantity <= spec.maxQuantity`. Because CRDs only
+  allow the `name` and `generatedName` to be declared in the `metadata` of an object, these are the
+  only metadata fields that may be validated using CEL validator rules. For example,
+  `metadata.name.endsWith('mySuffix')` is allowed, but only when the OpenAPIv3 schema explicitly
+  declares `metadata` as a field in the root object and declares the `name` field within `metadata`.
+  `size(metadata.labels) < 3` however, it not allowed. The limit on which `metadata` fields may be
+  validated is an intentional design choice (that aims to keep metadata behavior uniform across
+  types) and applies to all validation mechanisms (e.g. the OpenAPIV3 `maxItems` restriction), not
+  just CEL validator rules.
 
```

Comment:
I've confirmed with @TristonianJones that macro and function identifiers are in a separate namespace and cannot collide with variables or field names. I have tests for a variety of these cases that are passing.

Reply:
@liggitt in which direction are you concerned about conflicts? Are you thinking of global functions, member functions, both? The type checker will hard fail when there are signature collisions between functions so if you're using the standard environment and the extensions, and then try to add custom functions with overlapping signatures the environment setup itself would error.

Adding new macros is tricky because macro expansion happens before type-checking, which means you can have an overlapping signature between a macro and function, and the macro would always win. You can consider this a bug or a feature, but usually the expectation is that macros are used sparingly in comparison to functions and only when you absolutely need to access a feature that the syntax doesn't permit you to directly (comprehensions, 'test_only' expressions).

Reply:
I was mostly thinking of conflicts between identifiers and function/macro names. If the namespaces are separate, and we're limiting the top-level identifier to `self`, my concern is resolved.